### PR TITLE
BUG: Array ufunc reduce out tuple

### LINKED ISSUE #9102: np.max() doesn't behave properly for ndarray subclasses that define __array_ufunc__
Take the following example:

```
import numpy as np

class MyArr(np.ndarray):
    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        func = getattr(ufunc, method)
        inp = [i.view(np.ndarray) for i in inputs]
        return func(*inp, **kwargs)

data = np.array([[1, 2, 3], [1, 2, 3]])

arr = data.view(MyArr)

print(data.max())
print(arr.max())
```

On numpy 1.13rc1 this prints:

```
3
[1 2 3]
```

For the call to `arr.max()` numpy calls `__array_ufunc__`, but there doesn't seem to be a way to tell from inside `__array_ufunc__` that we should be reducing across the whole array and not just the first dimension. I think this would work properly if numpy were calling `__array_ufunc__` with `axis=None`.

Ping @mhvk

### PR DESCRIPTION
**Pull request by mhvk**
This ensures that the normal `reduce`, `accumulate` and `reduceat` methods can deal with getting `out` as a single-item tuple, which is needed to work with `__array_ufunc__` (see #9105). It builds on #9104; can be done together or in sequence.

note: I think I'm right that with my refcounting (not doing anything), but someone who understands this better please check...

Closes #9102.

### CODE REVIEW

**Comment by mhvk** [comment]
In case it isn't obvious, this bug also hits things that use `.reduce` in the background: without this, `a.max(out=b)` doesn't work either with `__array_ufunc__`.

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -3970,17 +3971,26 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
...
                                         &keepdims)) {
             Py_XDECREF(otype);
             return NULL;
         }
     }
+    /* if out_obj is a tuple with size 1 (=nout here), unpack it */
+    if (out_obj != NULL && PyTuple_CheckExact(out_obj) &&
+            PyTuple_GET_SIZE(out_obj) == 1) {
+        out_obj = PyTuple_GET_ITEM(out_obj, 0);
+    }
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Would be nice to give an explicit error here if the output is a tuple of the wrong length - elsewhere, we have:
```
if (PyTuple_GET_SIZE(value) != nout) {
    PyErr_SetString(PyExc_ValueError,
            "The 'out' tuple must have exactly "
            "one entry per ufunc output");
    goto fail;
}
```

File: numpy/core/tests/test_umath.py [hunk 2]
```diff
@@ -2041,6 +2048,8 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
                     return NotImplemented
 
                 if method == 'at':
+                    if isinstance(inputs[0], A):
+                        inputs[0].info = info
```
Comment (eric-wieser) [inline numpy/core/tests/test_umath.py]: Theres some documentation that should be kept in sync with this, isn't there?

Comment (mhvk) [inline numpy/core/tests/test_umath.py]: I *think* this PR makes the code consistent with the documentation... Unless you think we should explicitly mention in the reduce, accumulate, reduceat documentation that a tuple of 1 is allowed?

Comment (eric-wieser) [inline numpy/core/tests/test_umath.py]: I was thinking about the `__array_ufunc__` example in the docs, which I thought was kept synced with one of the tests. Maybe it was not that one.

Yes, I think we should document that a tuple is allowed for `out` in those functions too.

File: numpy/core/src/umath/ufunc_object.c (hunk 1 above)
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: OK, done, except a `TypeError`, as in `override.c`, because the situation here seems most analogous to not given sufficient number of arguments to a function, in which case cpython also raises a `TypeError` (see first item of the post-merger to-do list in #8892).

Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: I disagree - this is analogous to having the wrong number of outputs to a function, ie tuple unpacking - so `ValueError` is correct

**Comment by ngoldbaum** [comment]
Just a comment: what does it mean for `out` to be anything but a single array? In other words, in what context would you pass in a tuple? Maybe for ufuncs like `divmod` that return tuples? It would help a lot to discuss this in the `__array_ufunc__` docs. It might already be there and I'm missing it though.

**Comment by mhvk** [comment]
@ngoldbaum - this is really just for consistency with the regular `__call__` method, to ensure that `__array_ufunc__` always gets the same `out` keyword argument (just like `*inputs` is a tuple even if there is only one input). And who knows, maybe someone at some point can think of a function with two inputs and two outputs other than `divmod` where repeated application makes sense....

Anyway, given your comment, I should clearly try to make it more explicit in the `__array_ufunc__` documentation.

**Comment by eric-wieser** [comment]
@ngoldbaum: Functions like `divmod` need multiple output arguments. As such, `ufunc.__call__` accepts `out` as a tuple, with an array of each argument.

For simplicity of overrides, `out=arr` is promoted to `out=(arr,)` on entry to `__array_ufunc__`, so that both cases can be handled in the same way.

**Comment by eric-wieser** [comment]
@mhvk: We should only allow a tuple to be passed by keyword, not position, for consistency with `ufunc.__call__`

**Comment by mhvk** [comment]
OK, so the test should be higher up...  Will do. And will change to `ValueError` here and in `override.c`.

**Comment by mhvk** [comment]
OK, made the changes, including docs, and did the rebase. Haven't had much time to check whether things are all OK (though tests pass); have to run now, back in ~2 hours.

**Comment by charris** [comment]
Any reason this should not go into 1.13? I'm going to do another release in a few days to take care the current crop of errors.

File: numpy/add_newdocs.py [hunk 3]
```diff
@@ -5446,7 +5446,9 @@ def luf(lamdaexpr, *args, **kwargs):
         Input arrays.
     out : ndarray or tuple of ndarray, optional
         Alternate array object(s) in which to put the result; if provided, it
-        must have a shape that the inputs broadcast to.
+        must have a shape that the inputs broadcast to. A tuple of arrays
+        (possible only as a keyword argument) must have length equal to the
+        number of outputs; unwanted outputs can be set to `None`.
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: `None` doesn't mean "I don't want an output", just "you can allocate it yourself"

**Comment by mhvk** [comment]
@eric-wieser - I think all is OK now. Thanks for the rapid and careful review!

**Comment by mhvk** [comment]
p.s. Please check my logic for not touching any reference count in the additional check out `out`...

File: numpy/add_newdocs.py [hunk 4]
```diff
@@ -5822,7 +5828,9 @@ def luf(lamdaexpr, *args, **kwargs):
         the data type of the input array if no output array is provided.
     out : ndarray, optional
         A location into which the result is stored. If not provided a
-        freshly-allocated array is returned.
+        freshly-allocated array is returned. A tuple of arrays
+        (possible only as a keyword argument) must have length equal to the
+        number of outputs; use `None` for outputs to be allocated by the ufunc.
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: This is a bit wordy - can you combine "use `None` for outputs to be allocated by the ufunc" with "If not provided a freshly-allocated array is returned. "? Not provided and none are the same thing with the default documented as it is.

File: numpy/core/src/umath/ufunc_object.c [hunk 5]
```diff
@@ -3927,7 +3928,20 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
                      _reduce_type[operation]);
         return NULL;
     }
-
+    /* if there is a tuple of 1 for `out` in kwds, unpack it */
+    if (kwds != NULL) {
+        PyObject *out_obj = PyDict_GetItem(kwds, npy_um_str_out);
+        if (out_obj != NULL && PyTuple_Check(out_obj)) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Do we want `CheckExact`? Note that `PyTuple_GET_SIZE` can fail for tuple subclasses

**Review by eric-wieser — COMMENTED** [review]
Haven't checked the tests, but looks OK for refcounting

File: numpy/core/src/umath/ufunc_object.c (hunk 5 above)
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: OK, I'll change that, here and in the two other locations where `out` is treated exactly in the same way (once in this file, once in `override.c`).

File: numpy/add_newdocs.py (hunk 4 above)
Comment (mhvk) [inline numpy/add_newdocs.py]: How about
```
        A location into which the result is stored. If not provided or `None`,
        a freshly-allocated array is returned. A tuple (possible only as a
        keyword argument) must have length equal to the number of outputs.
```

**Comment by mhvk** [comment]
All done?

**Review by eric-wieser — COMMENTED** [review]
Generally LGTM. One last comment - I think we need a `.. versionchanged:: 1.13.0` for the tuple version of out for `reduce` and friends, and possibly a release note

**Comment by mhvk** [comment]
I'm wondering if we really want to have a `versionchanged` line for every ufunc. This is there just so we can deal with `__array_ufunc__` better; it seems overkill for something I would not recommend doing (and where do we stop -- with this PR, it is possible to do `a.max(out=(b,))`, but should we adjust all those docstrings too?).

**Comment by charris** [comment]
The parameter entries are generated, so if you add it to one, you will get it for all whether you wish it or not ;) You can always say it is not recommended, but I don't think it makes much difference.

**Comment by charris** [comment]
I think a `.. versionchanged::` would be appropriate for the three affected methods in `add_newdocs.py`. As the "new" functionality already worked for `__call__` in 1.12, I don't think it is needed for that. This is almost a bug fix in that regard. Something in the release notes would also be good if this is not already covered.

**Comment by eric-wieser** [comment]
> if we really want to have a versionchanged line for every ufunc.

The only documentation that needs to change is for `ufunc.reduce` etc, not for each function like `np.add.reduce` - there are no such documentation pages

File: doc/source/reference/ufuncs.rst [hunk 6]
```diff
@@ -426,8 +426,8 @@ Methods
 All ufuncs have four methods. However, these methods only make sense on
 ufuncs that take two input arguments and return one output argument.
 Attempting to call these methods on other ufuncs will cause a
-:exc:`ValueError`. The reduce-like methods all take an *axis* keyword
-and a *dtype* keyword, and the arrays must all have dimension >= 1.
+:exc:`ValueError`. The reduce-like methods all take an *axis* keyword, a *dtype*
+keyword, and an *out* keyword, and the arrays must all have dimension >= 1.
 The *axis* keyword specifies the axis of the array over which the reduction
 will take place and may be negative, but must be an integer. The
```
Comment (charris) [inline doc/source/reference/ufuncs.rst]: Hmm, this is outdated I think, tuples of ints also work. Not a problem for this, but needs fixing at some point. I'll open an issue.

**Comment by charris** [comment]
@eric-wieser Look good to you?

File: numpy/add_newdocs.py [hunk 7]
```diff
@@ -5741,9 +5747,13 @@ def luf(lamdaexpr, *args, **kwargs):
         The data-type used to represent the intermediate results. Defaults
         to the data-type of the output array if such is provided, or the
         the data-type of the input array if no output array is provided.
-    out : ndarray, optional
-        A location into which the result is stored. If not provided a
-        freshly-allocated array is returned.
+    out : ndarray, None, or tuple of ndarray and None, optional
+        A location into which the result is stored. If not provided or `None`,
+        a freshly-allocated array is returned. A tuple (possible only as a
+        keyword argument) must have length equal to the number of outputs.
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: I think it should be made obvious here that the number of outputs is always 1 for `reduce` etc

**Comment by ngoldbaum** [comment]
Is there going to be a numpy 1.13rc2? If not, I'd like to ask for one. I've been waiting on this going in to further experiment with `__array_ufunc__`.

File: numpy/add_newdocs.py (hunk 7 above)
Comment (mhvk) [inline numpy/add_newdocs.py]: done

**Comment by mhvk** [comment]
@ngoldbaum - yes, I agree we should have a second release candidate. @charris?

File: numpy/add_newdocs.py (hunk 7 above)
Comment (eric-wieser) [inline numpy/add_newdocs.py]: "all ufuncs implemented in numpy" doesn't cover it. This is a property of `ufunc.reduce`, so even third party ufuncs will conform to this.

I'd probably just go with `for consistency with `ufunc.__call__`, a tuple of length 1 is also allowed`.

Comment (mhvk) [inline numpy/add_newdocs.py]: To avoid iterating yet more on a change I didn't feel was really necessary to make (as it is just an implementation detail to help `__array_ufunc__`), could you give the exact text that you suggest?

**Comment by charris** [comment]
@mhvk Planning on it and would like to get it out this week. I started with rc1 because few look at the betas ...

File: numpy/add_newdocs.py (hunk 7 above)
Comment (eric-wieser) [inline numpy/add_newdocs.py]: Good point - iterating here is far easier than iterating and rebasing. Apologies for not thinking of that earlier. 

How about, for `reduce`, `at`, and `reduceat`:

```
out : ndarray, None, or tuple of ndarray and None, optional
    A location into which the result is stored. If not provided or `None`,
    a freshly-allocated array is returned. For consistency with :ref:`ufunc.__call__`,
    this may be wrapped in a 1-element tuple.
```

Comment (mhvk) [inline numpy/add_newdocs.py]: OK, thanks. Now I didn't follow this pattern, but did make a change, to add "if given as a keyword" (since we don't check for it if given as a positional element)

Comment (eric-wieser) [inline numpy/add_newdocs.py]: Yep, my mistake on forgetting that

File: numpy/core/src/umath/override.c [hunk 8]
```diff
@@ -360,11 +366,11 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
         if (out != NULL) {
             int nout = ufunc->nout;
 
-            if (PyTuple_Check(out)) {
+            if (PyTuple_CheckExact(out)) {
                 int all_none = 1;
 
                 if (PyTuple_GET_SIZE(out) != nout) {
-                    PyErr_Format(PyExc_TypeError,
+                    PyErr_Format(PyExc_ValueError,
```
Comment (eric-wieser) [inline numpy/core/src/umath/override.c]: This needs a release note - we already have a section for changed exceptions

Comment (mhvk) [inline numpy/core/src/umath/override.c]: This one has actually not been released yet (it is used in `__array_ufunc__`), and it now becomes consistent with the regular ufuncs, so I think it is OK.

Comment (eric-wieser) [inline numpy/core/src/umath/override.c]: You're right - testing `ufunc.__call__` on 1.12 gives me `ValueError`, as you have here.

**Review by eric-wieser — APPROVED** [review]
LGTM

**Comment by eric-wieser** [comment]
Will merge tomorrow, along with the backport PR I set up at #9111
